### PR TITLE
[module-minifier-plugin] Async import compression

### PR DESCRIPTION
**Pull request by dmichon-msft**
## Summary
Adds an option `compressAsyncImports: true` to `ModuleMinifierPlugin` that, if specified, reduces the API call size of webpack async import expressions by combining all chunk imports and the final module import into a single call to the webpack runtime. Also creates a lookup table of frequently used chunk sets to further reduce code size.

## Details
Converts expressions like `Promise.resolve([__webpack_require__.e(1), __webpack_require__.e(2)]).then(__webpack_require__.bind(23))` to the more concise `__webpack_require__.ee([1,2 ], 23)`

## How it was tested
Build test project `localization-plugin-test-02`
Linking inside a private repo with a large quantity of async imports.

### CODE REVIEW

File: webpack/module-minifier-plugin/src/PortableMinifierIdsPlugin.ts [hunk 1]
```diff
@@ -28,7 +29,7 @@ const TAP_AFTER: TapOptions<'sync'> = {
 };
 
 const STABLE_MODULE_ID_PREFIX: '__MODULEID_SHA_' = '__MODULEID_SHA_';
-const STABLE_MODULE_ID_REGEX: RegExp = /['"]?(__MODULEID_SHA_[0-9a-f]+)['"]?/g;
+const STABLE_MODULE_ID_REGEX: RegExp = /(?<!C)['"]?(__MODULEID_SHA_[0-9a-f]+)['"]?/g;
```
Comment (iclanton) [inline webpack/module-minifier-plugin/src/PortableMinifierIdsPlugin.ts:32]: Can you leave a comment describing what this matches?

File: webpack/module-minifier-plugin/src/PortableMinifierIdsPlugin.ts [hunk 2]
```diff
@@ -65,29 +66,36 @@ export class PortableMinifierModuleIdsPlugin implements Plugin {
 
     const stableIdToFinalId: Map<string | number, string | number> = new Map();
 
-    this._minifierHooks.finalModuleId.tap(PLUGIN_NAME, (id: string | number | undefined) => {
+    const { finalModuleId } = this._minifierHooks;
+
+    finalModuleId.tap(PLUGIN_NAME, (id: string | number | undefined) => {
```
Comment (iclanton) [inline webpack/module-minifier-plugin/src/PortableMinifierIdsPlugin.ts:71]: This is less readable.

File: webpack/module-minifier-plugin/src/AsyncImportCompressionPlugin.ts [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+  module: webpack.compilation.Module,
+  originModule: webpack.compilation.Module
+): string {
+  const exportsType: string | undefined = module.buildMeta?.exportsType;
+  const strict: boolean | undefined = originModule.buildMeta?.strictHarmonyModule;
+
+  if (exportsType === 'namespace') {
+    return '';
+  } else if (exportsType === 'named') {
+    return ',3';
```
Comment (iclanton) [inline webpack/module-minifier-plugin/src/AsyncImportCompressionPlugin.ts:67]: Where do these numbers come from? Can you stick them in consts with meaningful names?

File: webpack/module-minifier-plugin/src/AsyncImportCompressionPlugin.ts [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+  for (const chunkGroup of chunk.groupsIterable) {
+    if (chunkGroup.getNumberOfChildren() > 0) {
+      return true;
+    }
+  }
+  return false;
+}
+
+/**
+ * Plugin that replaces `Promise.all([__webpack_require__.e(1), __webpack_require__,e(12)]).then(__webpack_require__.t.bind(123,7))`
```
Comment (iclanton) [inline webpack/module-minifier-plugin/src/AsyncImportCompressionPlugin.ts:85]:
```suggestion
 * Plugin that replaces `Promise.all([__webpack_require__.e(1), __webpack_require__.e(12)]).then(__webpack_require__.t.bind(123,7))`
```

**Review by iclanton — APPROVED** [review]

File: webpack/module-minifier-plugin/src/AsyncImportCompressionPlugin.ts (hunk 3 above)
Comment (dmichon-msft) [reply webpack/module-minifier-plugin/src/AsyncImportCompressionPlugin.ts:67]: They come from webpack's source code. I'll add some comments or at least a reference.

**Pull request merged**